sheet08/sheet08solutions.ipynb

```diff
@@ -29,294 +29,231 @@
+    "#### Back-Propagation / Adaptation\n",
+    "In order to learn something the perceptron needs to slowly adjust its weight according to its provided error signal delta (which will be calculated by the MultilayerPerceptron). Each weight $w_{j,i}$ in layer $k$ is adjusted by a value $\\Delta w_{j,i}$ according to the following formula given a learning rate $\\epsilon$:\n",
     "\n",
-    "The second cell is a simple plot to give you an idea of how this looks like. By now you probably already figured out whether this task can be done with a simple perceptron or not. If you didn't, run the first and second cell and take a closer look at the data."
+    "$$\\Delta w_{j,i}(k) = \\epsilon\\, \\delta_j(k+1)\\, o_i(k)$$"
    ]
   },
   {
    "cell_type": "code",
    "execution_count": null,
-   "metadata": {
-    "collapsed": false
-   },
+   "metadata": {},
    "outputs": [],
    "source": [
     "import numpy as np\n",
     "\n",
-    "# Generate some data.\n",
-    "ND = 1000\n",
-    "input_dim = 3\n",
-    "D = np.random.rand(ND, input_dim)\n",
-    "# Label data: sum should be > 0.7 * dim.\n",
-    "T = (np.sum(D, 1) > 0.7 * input_dim) * 1"
-   ]
-  },
-  {
-   "cell_type": "code",
-   "execution_count": null,
-   "metadata": {
-    "collapsed": false,
-    "scrolled": false
-   },
-   "outputs": [],
-   "source": [
-    "%matplotlib notebook\n",
-    "import matplotlib.pyplot as plt\n",
-    "from mpl_toolkits.mplot3d import Axes3D\n",
+    "# Activation function \\sigma.\n",
+    "# We use scipy's builtin because it fixes some NaN problems for us.\n",
+    "# sigmoid = lambda x: 1 / (1 + np.exp(-x))\n",
+    "from scipy.special import expit as sigmoid\n",
+    "\n",
+    "\n",
+    "class Perceptron:\n",
+    "    \"\"\"Single neuron handling its own weights and bias.\"\"\"\n",
+    "\n",
+    "    def __init__(self, dim_in, act_func=sigmoid):\n",
+    "        \"\"\"Initialize a new neuron with its weights and bias.\n",
+    "\n",
+    "        Args:\n",
+    "            dim_in: Dimensionality of the data coming into this perceptron. \n",
+    "                In a network of perceptrons this basically represents the \n",
+    "                number of neurons in the layer before this neuron's layer. \n",
+    "                Used for generating the perceptron's weights vector, which \n",
+    "                not only includes one weight per input but also an additional \n",
+    "                bias weight.\n",
+    "            act_fun: Function to apply on activation.\n",
+    "        \"\"\"\n",
+    "        self.act_func = act_func\n",
+    "        # self.weights = np.array([])  # TODO\n",
+    "        self.weights = np.random.normal(size=dim_in + 1)\n",
```

Comment:
The lecture apparently does not explicitly talk about biases. We implement them using an additional weight, which is kind of intuitive but simplifies applying it to the inputs greatly. Might need further explanations, maybe in the practice session?